ethToElrond/v2/ethToElrondBridgeExecutor.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	return &ethToElrondBridgeExecutor{
+		log:              args.Log,
+		topologyProvider: args.TopologyProvider,
+		elrondClient:     args.ElrondClient,
+		ethereumClient:   args.EthereumClient,
+	}, nil
+}
+
+// GetLogger returns the logger implementation
+func (executor *ethToElrondBridgeExecutor) GetLogger() logger.Logger {
+	return executor.log
+}
+
+// MyTurnAsLeader returns true if the current relayer node is the leader
+func (executor *ethToElrondBridgeExecutor) MyTurnAsLeader() bool {
+	return executor.topologyProvider.MyTurnAsLeader()
+}
+
+// GetAndStoreActionID fetches the action ID by using the stored batch and will return and store the action ID
+func (executor *ethToElrondBridgeExecutor) GetAndStoreActionID(ctx context.Context) (uint64, error) {
+	if executor.batch == nil {
+		return 0, errNilBatch
+	}
+
+	actionID, err := executor.elrondClient.GetActionIDForProposeTransfer(ctx, executor.batch)
+	if err != nil {
+		return 0, err
+	}
+
+	executor.actionID = actionID
+
+	return actionID, nil
+}
+
+// GetStoredActionID will return the stored action ID
+func (executor *ethToElrondBridgeExecutor) GetStoredActionID() uint64 {
+	return executor.actionID
+}
+
+// GetAndStoreBatchFromEthereum will fetch and store the batch from the ethereum client
+func (executor *ethToElrondBridgeExecutor) GetAndStoreBatchFromEthereum(ctx context.Context, nonce uint64) error {
+	batch, err := executor.ethereumClient.GetBatch(ctx, nonce)
+	// TODO add error filtering here
+	if err != nil {
+		return err
+	}
+
+	executor.batch = batch
+
+	return nil
+}
+
+// GetStoredBatch returns the stored batch
+func (executor *ethToElrondBridgeExecutor) GetStoredBatch() *clients.TransferBatch {
+	return executor.batch
+}
+
+// GetLastExecutedEthBatchIDFromElrond returns the last executed batch ID that is stored on the Elrond SC
+func (executor *ethToElrondBridgeExecutor) GetLastExecutedEthBatchIDFromElrond(ctx context.Context) (uint64, error) {
+	return executor.elrondClient.GetLastExecutedEthBatchID(ctx)
+}
+
+// VerifyLastDepositNonceExecutedOnEthereumBatch will check the deposit nonces from the fetched batch from Ethereum client
+func (executor *ethToElrondBridgeExecutor) VerifyLastDepositNonceExecutedOnEthereumBatch(ctx context.Context) error {
+	if executor.batch == nil {
+		return errNilBatch
+	}
+
+	lastNonce, err := executor.elrondClient.GetLastExecutedEthTxID(ctx)
+	if err != nil {
+		return err
+	}
+
+	return executor.verifyDepositNonces(lastNonce)
+}
+
+func (executor *ethToElrondBridgeExecutor) verifyDepositNonces(lastNonce uint64) error {
+	startNonce := lastNonce + 1
+	for _, dt := range executor.batch.Deposits {
+		if dt.Nonce != startNonce {
+			return fmt.Errorf("%w for deposit %s", errInvalidDepositNonce, dt.String())
+		}
+
+		startNonce++
+	}
+
+	return nil
+}
+
+// WasTransferProposedOnElrond checks if the transfer was proposed on Elrond
+func (executor *ethToElrondBridgeExecutor) WasTransferProposedOnElrond(ctx context.Context) (bool, error) {
```

Comment:
Just a comment, nothing to be done in this PR: Isn't it a bit tricky to return bool + error, since they both mean the same thing? In case of error, bool will always be false and vice-versa. 
Perhaps:
- only return error and compare it to nil
- only return bool and log errors

Reply:
Not quite tricky. We can have the following situations:
```
(true, nil) // the transfer was done, no error during fetch operation
(false, nil) // the transfer was not done, no error during fetch operation
(false, error) //unable to tell if the transfer was done or not because the fetch operation failed with error
```

Reply:
Right, good point